src/Bicep.Core/Semantics/Namespaces/SystemNamespaceSymbol.cs

```diff
@@ -520,6 +511,33 @@ private static IEnumerable<Decorator> GetSystemDecorators()
                 .WithEvaluator(MergeToTargetObject("metadata", decoratorSyntax => SyntaxFactory.CreateObject(
                     SyntaxFactory.CreateObjectProperty("description", SingleArgumentSelector(decoratorSyntax)).AsEnumerable())))
                 .Build();
+
+            yield return new DecoratorBuilder("batchSize")
+                .WithDescription("Causes the resource or module for-expression to be run in sequential batches of specified size instead of the default behavior where all the resources or modules are deployed in parallel.")
+                .WithRequiredParameter("batchSize", LanguageConstants.Int, "The size of the batch")
+                .WithFlags(FunctionFlags.ResourceOrModuleDecorator)
+                // the decorator is constrained to resources and modules already - checking for array alone is simple and should be sufficient
+                .WithValidator((decoratorName, decoratorSyntax, targetType, typeManager, diagnosticWriter) =>
+                {
+                    if (!TypeValidator.AreTypesAssignable(targetType, LanguageConstants.Array))
+                    {
+                        // the resource/module declaration is not a collection
+                        // (the compile-time constnat and resource/module placement is already enforced, so we don't need a deeper type check)
+                        diagnosticWriter.Write(DiagnosticBuilder.ForPosition(decoratorSyntax).BatchSizeNotAllowed(decoratorName));
+                    }
+
+                    // do not rely on casting - the type checker does not guarantee IntegerLiteralSyntax
+                    if (SingleArgumentSelector(decoratorSyntax) is IntegerLiteralSyntax batchSizeSyntax)
+                    {
+                        const long MinimumBatchSize = 1;
+                        if (batchSizeSyntax.Value < MinimumBatchSize)
```

Comment:
>MinimumBatchSize [](start = 52, length = 16)

@shenglol we should probably add min/max checks to other decorators that take integer arguments at some point.

Reply:
Yeah makes sense. I create https://github.com/Azure/bicep/issues/1730 to track it.